frontend/apps/reader/modules/readerhighlight.lua

```diff
@@ -297,6 +306,43 @@ end
 
 function ReaderHighlight:onReaderReady()
     self:setupTouchZones()
+    self:migrateHighlights()
+end
+
+
+function ReaderHighlight:migrateHighlights()
+    if self.ui.doc_settings:hasNot("highlight_merged") then
+        for page, highlights in pairs(self.view.highlight.saved) do
+            for _, item in ipairs(highlights) do
+                -- This is to switch notes and text.
+                local highlight = bookmarkToFromLocalHighlight(item)
+                if not highlight.chapter then
+                    local pg_or_xp = self.ui.paging and page or highlight.pos0
+                    highlight.chapter = self.ui.toc:getTocTitleByPage(pg_or_xp)
+                end
+                if not highlight.drawer then
+                    highlight.drawer = self.view.highlight.saved_drawer
+                end
+                highlight.page_num = page
+                if self.ui.paging then
+                    highlight.page = page
+                end
+                self.ui.bookmark:patchBookmark({
+                    datetime = highlight.datetime,
+                    page = highlight.page,
+                    updated_highlight = highlight
+                });
+            end
+        end
+        for i, bookmark in ipairs(self.ui.bookmark.bookmarks) do
+            if bookmark.highlighted then
+                if not bookmark.drawer then
+                    self.ui.bookmark.bookmarks[i].drawer = self.view.highlight.saved_drawer
+                end
+            end
+        end
+    end
+    self.view.highlight.saved = nil;
```

Comment:
I think we should at least keep a backup of highlights in the doc settings (so it's available for manual fixing if mess happened).
Or even a backup of the metadata.xxx.lua itself. I do something like that in:
https://github.com/koreader/koreader/blob/f37699a12b893984690f0e433b7f048af3c621a7/frontend/apps/reader/modules/readerrolling.lua#L1336-L1344

Might also want to see how the various migrations we do happen vs. this one. Your should happen after all the others.
Dunno (I haven't really thought about it, all this is very old stuff :) if doing it in `ReaderHighlight:onReaderReady()` is the right moment - seems the cre_dom_version would be done after in a postReaderCallback:
https://github.com/koreader/koreader/blob/f37699a12b893984690f0e433b7f048af3c621a7/frontend/apps/reader/readerui.lua#L458-L465

Reply:
At least, the old `highlight` table can be renamed to `highlight_obsolete` (only previously opened book will keep it) so that when testing old koreader versions for regression/performance, we can just manually rename this key in the settings.

Reply:
I'm saving the previous highlights in `obsolate_highlights` in settings. That should do, right? Or, should I backup the metadata file?

Reply:
Should do. But keep the singular/typo `highlight` (and I personally prefer `highlight_obsolete` with _obsolete at the end, easier to see/read/restore).
And even `highlight_rolling_obsolete` / `highlight_paging_obsolete` because of the thing I mentionned in some other comment, as we may have both (very rare, but possible).

(I'm off for a week, I won't be able to read/comment as quick/much as usually.)